### PR TITLE
Write data parameter docs as regular parameter not as note

### PR DESCRIPTION
**Pull request by timhoffm**
## PR Summary

It's cleaner to document this as a normal parameter rather than using a note "The function can take an additional parameter *data* ...".

The docs is inserted before `**kwargs` docs, or if they do not exist as the last entry in `Parameters` (Which is the case only for a handful of methods.

### CODE REVIEW

File: lib/matplotlib/__init__.py [hunk 1]
```diff
@@ -1247,20 +1247,12 @@ def _label_from_arg(y, default_name):
...
-"""
-
-_DATA_DOC_APPENDIX = """
-
-.. note::
-    In addition to the above described arguments, this function can take
-    a *data* keyword argument. If such a *data* argument is given,
+_DATA_KWARG = """\
+data : indexable object, optional
+    If given,
```
Comment (anntzer) [inline lib/matplotlib/__init__.py:1252]: Move that "if given" to the beginning of `replaced`, to avoid a slightly strange linebreak in the unrendered docs?

Comment (timhoffm) [reply lib/matplotlib/__init__.py]: I've moved this into the formatted message. No need to split the sentence.

File: lib/matplotlib/__init__.py [hunk 2]
```diff
@@ -1287,17 +1278,45 @@ def _add_data_doc(docstring, replace_names):
...
+        if insert_index > 0:
+            return insert_index
+        # If **kwargs docs does not exist, insert as the last parameter.
+        # To do so, find the end of the parameters section:
+        # - find start of parameters section
+        # - find next section (defined by a line starting with '----'). This
+        #   seems the most reliable way as the parameters section may be quite
+        #   complex.
+        # - Go back to the previous empty line (A section title must be
+        #   preceededby an emtpy line.
```
Comment (QuLogic) [inline lib/matplotlib/__init__.py:1294]:
```suggestion
        #   preceeded by an emtpy line.
```

File: lib/matplotlib/tests/test_preprocess_data.py [hunk 3]
```diff
@@ -194,35 +194,105 @@ def func(ax, x, y, z=1):
...
+        Funcy does nothing.
+
+        Parameters
+        ----------
+        **kwargs
+            Text.
+        """
+    assert not re.search(r"all parameters also accept a string", funcy.__doc__)
+    assert not re.search(r"the following parameters .*: \*bar\*\.",
                          funcy.__doc__)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_preprocess_data.py:282]:
```suggestion
    assert re.search(r"the following parameters .*:\n    \*bar\*\.",
                     funcy.__doc__, re.DOTALL)
```

File: lib/matplotlib/tests/test_preprocess_data.py [hunk 4]
```diff
@@ -194,35 +194,105 @@ def func(ax, x, y, z=1):
...
+        Funcy does nothing.
+
+        Parameters
+        ----------
+        x, y, z
+            Text.
+        """
+    assert not re.search(r"all parameters also accept a string", funcy.__doc__)
+    assert not re.search(r"the following parameters .*: \*x\*, \*t\*\.",
                          funcy.__doc__)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_preprocess_data.py:296]:
```suggestion
    assert re.search(r"the following parameters .*:\n    \*x\*, \*t\*\.",
                     funcy.__doc__, re.DOTALL)
```

**Review by QuLogic — COMMENTED** [review]
@tacaswell and I were looking at this earlier, and realized that a couple of the asserts in tests with `replace_names` were wrong. That text is there, but the regex is wrong.

**Comment by QuLogic** [comment]
I'm not completely sure whether this belongs in 3.4.2; I'll let you know after docs have built.

**Comment by timhoffm** [comment]
I'm fine with moving to 3.5 if this would hold up the 3.4.2 release.

**Comment by QuLogic** [comment]
If `**kwargs* is in 'Other Parameters`, then `data` ends up there; is that what was intended?

**Comment by timhoffm** [comment]
> If `**kwargs*` is in 'Other Parameters', then `data` ends up there; is that what was intended?

I don't have a strong opinion on this, but it's fine by me. One can argue this on a similar level as further Artist properties, reatively generic and not related to the primary semantic of the method.

**Comment by jklymak** [comment]
Sorry for being dense - where is this change in the built docs?  I can't tell from the description.

**Comment by timhoffm** [comment]
It's removing the note
image

and instead adding a regular parameter:
image

This change applies to all `@_preprocess_data()` decorated methods, e.g.:
https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.bar.html?highlight=bar
vs.
https://58245-1385122-gh.circle-artifacts.com/0/doc/build/html/api/_as_gen/matplotlib.axes.Axes.bar.html?highlight=bar

**Comment by jklymak** [comment]
I see, I looked at `plot` where `data` is already explicitly described so wasn't sure what you are talking about.   However, is there a reason not to simply include `data` explicitly like `plot` does in the relevant methods?  I know thats kind of what this does, but rather than having a machinery, but machinery is mysterious unless you know it is there.

**Comment by timhoffm** [comment]
> However, is there a reason not to simply include `data` explicitly like `plot` does in the relevant methods?

### TL;DR

`_preprocess_data` does multiple things, that need to match. It's not strictily necessary to have all of them in the decorator.
My main motivation here is the final formatting of the docstring. I don't have a strong opinion how that is achieved internally.

### Full comment

There are three parts:

#### The docstring

We could make that explicit, but then might want to add a check that the mentioned parameters match with the given `@_preprocess_data` parameters.

An in-between way would be to
```
data : indexable object, optional
    {data_arg_placeholder}
```
so that we explicitly specify the position but still auto-generate the content.

#### The keyword `data`

Could be made explicit in the signature instead of the decorator rewriting the signature.

#### The actual data replacement
Also this does not necessarily have to be a decorator. You could do
```
if data is not None:
    x, y = replace_data(x, y, data)
```
Well, actually it's a bit more involved in general.

**Comment by timhoffm** [comment]
Thinking of it, I'd go for `{data_arg_placeholder}` and making `data=None` to reduce the magic here.

**Comment by timhoffm** [comment]
Closing in favor of #20291.
